WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,350 @@
+				{
+					takeOneMore();
+				}
+			}
+
+			largestMagnitudeNode = (fanIn ? negative.First() : positive.First() );
+			smallMagnitudeNodes = (fanIn ? positive.Take(nPositive).Reverse() : negative.Take(nNegative)); // reverse positive values so we always proceed in order of increasing magnitude
+
+			return true;
+		}
+
+		// Build a k-ary tree bottom up to reduce a list of nodes.
+		private ImmutableList<RequestNode> LimitCount(IEnumerable<RequestNode> nodes, int maxCount, CredentialType credentialType)
+		{
+			var nodeQueue = nodes.ToImmutableList();
+
+			while (nodeQueue.Count > maxCount)
+			{
+				// Replace up to k nodes, possibly the entire queue, with a
+				// single reissuance node which combines their values. The total
+				// number of items might be less than K but still larger than
+				// maxCount.
+				var take = Math.Min(K, nodeQueue.Count);
+				var reissuance = NewReissuanceNode();
+				DrainReissuance(reissuance, nodeQueue.Take(take), credentialType);
+				nodeQueue = nodeQueue.RemoveRange(0, take).Add(reissuance);
+			}
+
+			return nodeQueue;
+		}
+
+		private RequestNode NewReissuanceNode()
+		{
+			var node = new RequestNode(Vertices.Count, Enumerable.Repeat(0L, K).ToImmutableArray());
+			AddNode(node);
+			return node;
+		}
+
+		// Drain values into a reissuance request (towards the center of the graph).
+		private void DrainReissuance(RequestNode x, IEnumerable<RequestNode> ys, CredentialType type)
+		{
+			// The nodes' initial balance determines edge direction. Reissuance
+			// nodes always have a 0 initial value.
+			var xIsSink = x.InitialBalance(type).CompareTo(ys.First().InitialBalance(type)) == -1;
+
+			foreach (var y in ys)
+			{
+				// The amount for the edge is always determined by the `y`
+				// values, since we only add reissuance nodes to reduce the
+				// number of values required.
+				long amount = Balance(y, type);
+
+				if (xIsSink)
+				{
+					AddEdge(new(y, x, type, (ulong)amount));
+				}
+				else
+				{
+					AddEdge(new(x, y, type, (ulong)(-1 * amount)));
+				}
+
+				// Also drain all of the other credential types, to minimize
+				// dependencies between different requests, weight
+				// credentials should often be easily satisfiable with
+				// parallel edges to the amount credential edges.
+				for (CredentialType extraType = type + 1; extraType < CredentialType.NumTypes; extraType++)
+				{
+					var extraBalance = Balance(y, extraType);
+					if (extraBalance != 0 )
+					{
+						if (xIsSink)
+						{
+							AddEdge(new(y, x, extraType, (ulong)extraBalance));
+						}
+						else
+						{
+							AddEdge(new(x, y, extraType, (ulong)(-1 * extraBalance)));
+						}
+					}
+				}
+			}
+		}
+
+		// Drain credential values between terminal nodes, cancelling out
+		// opposite values by propagating forwards or backwards corresponding to
+		// fan-in and fan-out dependency structure.
+		private void DrainTerminal(RequestNode x, IEnumerable<RequestNode> ys, CredentialType type)
```

Comment:
Same here.

Reply:
not sure what to name them here... `node` and `nodes`?